src/mgr/OSDPerfMetricTypes.cc

```diff
@@ -104,19 +77,23 @@ std::ostream& operator<<(std::ostream& os,
   case PerformanceCounterType::READ_LATENCY:
     return os << "read latency";
   default:
-    return os << "unknown (" << d.type << ")";
+    return os << "unknown (" << static_cast<int>(d.type) << ")";
   }
 }
 
-bool OSDPerfMetricQuery::get_key(const OpRequest& op, std::string *key) const {
-  const MOSDOp* const m = static_cast<const MOSDOp*>(op.get_req());
-
-  *key = stringify(m->get_reqid().name);
+bool OSDPerfMetricQuery::get_key(GetSubKeyFnc get_sub_key_fnc,
```

Comment:
Nit: perhaps move this to the header and swap the use of ```std::function``` for just a lambda?

Reply:
@dillaman Do you mean something like this?
```
  bool get_key(
      bool get_sub_key(const OSDPerfMetricSubKeyDescriptor &,
                       OSDPerfMetricSubKey *),
      OSDPerfMetricKey *key) const {    
    for (auto &sub_key_descriptor : key_descriptor) {
      OSDPerfMetricSubKey sub_key;
      if (!get_sub_key(sub_key_descriptor, &sub_key)) {
        return false;
      }
      key->push_back(sub_key);
    }
    return true;
  }
```
It does not work for lambdas that are using captures [1], which is my case.

[1] https://stackoverflow.com/questions/7852101/c-lambda-with-captures-as-a-function-pointer

Reply:
No, I was saying something like (untested):

```
template <typename L>
bool get_key(L&& get_sub_key, OSDPerfMetricKey *key) const {
    for (auto &sub_key_descriptor : key_descriptor) {
      OSDPerfMetricSubKey sub_key;
      if (!get_sub_key(sub_key_descriptor, &sub_key)) {
        return false;
      }
      key->push_back(sub_key);
    }
    return true;
}
```

There shouldn't be a need to use a function pointer or ```std::function``` if the definition is in the header and it's templated around the lambda.